File: src/planets/mergePlanetOwnership.ts
import type { Planet } from "../data/types/planet.ts";
import type { PlanetOwnershipEntry } from "../data/types/planetOwnershipEntry.ts";
import { DEFAULT_PLANET_OWNERSHIP_ENTRY } from "../data/types/planetOwnershipEntry.ts";

// Colonist-Driven Production (planet-ownership.md). Merges a persisted
// planetOwnershipState entry onto a freshly-generated Planet at read
// time -- the same "normalize the live-read value, never trust the
// regenerated object's own field" pattern getDiscoveredPlanets() already
// established for `discovered`. Pure function: the SaveSystem-backed
// lookup lives in the caller (src/presentation/planetOwnershipState.ts).
//
// Retroactive removal (2026-08-04): citadelLevel/ownedByPlayerId merging
// removed along with Citadels -- see planet-ownership.md's own
// retroactive note. colonistCount merging is unaffected.
export function mergePlanetOwnership(planet: Planet, entry: PlanetOwnershipEntry | undefined): Planet {
  const resolved = entry ?? DEFAULT_PLANET_OWNERSHIP_ENTRY;
  return {
    ...planet,
    colonistCount: resolved.colonistCount,
  };
}

File: unity/ProfitableCore/Simulation/MergePlanetOwnership.cs
using Profitable.Core.Schema;

namespace Profitable.Core.Simulation;

// Ports src/planets/mergePlanetOwnership.ts. Merges a persisted
// planetOwnershipState entry onto a freshly-generated Planet at read time
// -- the same "normalize the live-read value, never trust the
// regenerated object's own field" pattern already established for
// Discovered. Pure function: the SaveSystem-backed lookup lives in the
// caller (Unity's PlanetOwnershipState.cs).
//
// Retroactive removal (2026-08-04): CitadelLevel/OwnedByPlayerId merging
// removed along with Citadels -- see planet-ownership.md's own
// retroactive note. ColonistCount merging is unaffected.
public static class PlanetOwnershipMerger
{
    public static Planet MergePlanetOwnership(Planet planet, PlanetOwnershipEntry? entry)
    {
        var resolved = entry ?? PlanetOwnershipEntry.Default();
        return new Planet
        {
            Id = planet.Id,
            Name = planet.Name,
            ProducibleResourceIds = planet.ProducibleResourceIds,
            PlanetType = planet.PlanetType,
            Tier = planet.Tier,
            Position = planet.Position,
            SpecialtyResourceId = planet.SpecialtyResourceId,
            Discovered = planet.Discovered,
            ResourceQualities = planet.ResourceQualities,
            ColonistCount = resolved.ColonistCount,
        };
    }
}
